### PR TITLE
feat: support resume

### PR DESCRIPTION
**Pull request by cl-shaharshavit-rnd**

### CODE REVIEW

File: MergerLogic/DataTypes/Fs.cs [hunk 1]
```diff
@@ -7,6 +7,7 @@ public class FS : Data
     {
         private IEnumerator<Tile> tiles;
         private bool done;
+        private int complatedTiles;
```
Comment (shimoncohen) [inline MergerLogic/DataTypes/Fs.cs:10]: completedTiles

File: MergerLogic/DataTypes/Fs.cs [hunk 2]
```diff
@@ -22,6 +23,7 @@ public override void Reset()
             this.tiles = this.GetTiles();
             this.tiles.MoveNext();
             this.done = false;
+            this.complatedTiles = 0;
```
Comment (shimoncohen) [inline MergerLogic/DataTypes/Fs.cs:26]: completedTiles

File: MergerLogic/DataTypes/Fs.cs [hunk 3]
```diff
@@ -68,9 +72,22 @@ public override List<Tile> GetNextBatch()
...
 
+            this.complatedTiles += tiles.Count;
+
             return tiles;
         }
 
+        public override void setBatchIdentifier(string batchIdentifier)
+        {
+            this.complatedTiles = int.Parse(batchIdentifier);
+            //un commont to make this function work at any point of the run and not only after the source initialization
```
Comment (shimoncohen) [inline MergerLogic/DataTypes/Fs.cs:83]: *comment

**Review by shimoncohen — CHANGES_REQUESTED** [review]

File: MergerLogic/DataTypes/Fs.cs (hunk 3 above)
Comment (asafMasa) [inline MergerLogic/DataTypes/Fs.cs:83]: un commont - spelling

File: MergerCli/utils/BatchStatusManager.cs [hunk 4]
```diff
@@ -0,0 +1,78 @@
+﻿using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace MergerCli.Utils
+{
+    internal class BatchStatusManager
+    {
+        internal class LayerStatus
```
Comment (asafMasa) [inline MergerCli/utils/BatchStatusManager.cs:8]: i don't like the you define 2 classes at the same file.
maybe try to separate this class definition to a different file

File: MergerCli/Proccess.cs [hunk 5]
```diff
@@ -1,26 +1,40 @@
...
+            batchStatusManager.InitilaizeLayer(newData.path);
             List<Tile> tiles = new List<Tile>(batchSize);
             int totalTileCount = newData.TileCount();
             int tileProgressCount = 0;
 
+            string? resumeBatchIdentifier = batchStatusManager.GetLayerBatchIdentifier(newData.path);
+            if (resumeBatchIdentifier != null)
+            {
+                newData.setBatchIdentifier(resumeBatchIdentifier);
+                //fix resume progress bug for gpkg and fs, fixing it for s3 reuries storing additional data.
```
Comment (asafMasa) [inline MergerCli/Proccess.cs:22]: reuries - spelling - requires?

File: MergerCli/utils/BatchStatusManager.cs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+        public void InitilaizeLayer(string layer)
+        {
+            if (!this.States.ContainsKey(layer))
+            {
+                this.States.Add(layer, new LayerStatus());
+            }
+        }
+
+        public void ComplateLayer(string layer)
```
Comment (asafMasa) [inline MergerCli/utils/BatchStatusManager.cs:44]: spelling

**Review by asafMasa — CHANGES_REQUESTED** [review]

File: MergerCli/Program.cs [hunk 7]
```diff
@@ -28,17 +32,23 @@ private static void Main(string[] args)
                                     {programName} 1000 gpkg geo.gpkg gpkg area1.gpkg gpkg area2.gpkg gpkg area3.gpkg
                                     {programName} 1000 gpkg geo.gpkg s3 area1 s3 area2 s3 area3
                                     {programName} 1000 s3 geo gpkg area1 s3 area2 gpkg area3
+
+                                    Resume: {programName} <path to status file>.
+                                    status file named 'status.json' will be created in running directory on incomplate merges
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerCli/Program.cs:37]: fixed

File: MergerCli/Program.cs [hunk 8]
```diff
@@ -51,60 +61,111 @@ private static void Main(string[] args)
...
+                }
+                string json = File.ReadAllText(args[1]);
+                batchStatusManager = BatchStatusManager.FromJson(json);
+                args = batchStatusManager.Command;
+            }
+            else
+            {
+                batchStatusManager = new BatchStatusManager(args);
+            }
+            AssemblyLoadContext.Default.Unloading += delegate { OnFailure(); };
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerCli/Program.cs:153]: added

File: MergerCli/Program.cs [hunk 9]
```diff
@@ -51,60 +61,111 @@ private static void Main(string[] args)
...
+                    Stopwatch stopWatch = new Stopwatch();
                     stopWatch.Start();
 
-                    Console.WriteLine("Validating merged datasources");
-                    Proccess.Validate(baseData, newData);
-
+                    string newType = args[sourceIndex];
+                    string newPath = args[sourceIndex + 1];
+
+                    if (batchStatusManager.IsLayerComplated(newPath))
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerCli/Program.cs:77]: added

File: MergerCli/utils/BatchStatusManager.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerCli/utils/BatchStatusManager.cs:44]: fixed

File: MergerCli/utils/BatchStatusManager.cs [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+        {
+            this.States[layer].IsDone = true;
+        }
+
+        public bool IsLayerComplated(string layer)
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerCli/utils/BatchStatusManager.cs:49]: fixed

File: MergerCli/Program.cs [hunk 11]
```diff
@@ -28,17 +32,23 @@ private static void Main(string[] args)
                                     {programName} 1000 gpkg geo.gpkg gpkg area1.gpkg gpkg area2.gpkg gpkg area3.gpkg
                                     {programName} 1000 gpkg geo.gpkg s3 area1 s3 area2 s3 area3
                                     {programName} 1000 s3 geo gpkg area1 s3 area2 gpkg area3
+
+                                    Resume: {programName} <path to status file>.
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerCli/Program.cs:36]: ???

File: MergerLogic/DataTypes/Fs.cs (hunk 1 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/DataTypes/Fs.cs:10]: fixed

File: MergerLogic/DataTypes/Fs.cs (hunk 2 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/DataTypes/Fs.cs:26]: fixed

File: MergerLogic/DataTypes/Fs.cs (hunk 3 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/DataTypes/Fs.cs:83]: fixed

File: MergerCli/utils/BatchStatusManager.cs [hunk 12]
```diff
@@ -0,0 +1,78 @@
+﻿using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace MergerCli.Utils
+{
+    internal class BatchStatusManager
+    {
+        internal class LayerStatus
+        {
+            public string? BathIdentifier { get; set; }
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerCli/utils/BatchStatusManager.cs:10]: fixed

File: MergerLogic/DataTypes/Fs.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerLogic/DataTypes/Fs.cs:10]: fixed

File: MergerLogic/DataTypes/Fs.cs (hunk 3 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/DataTypes/Fs.cs:83]: fixed

File: MergerCli/Proccess.cs (hunk 5 above)
Comment (cl-shaharshavit-rnd) [reply MergerCli/Proccess.cs:22]: fixed

File: MergerCli/utils/BatchStatusManager.cs (hunk 6 above)
Comment (cl-shaharshavit-rnd) [reply MergerCli/utils/BatchStatusManager.cs:44]: fixed

File: MergerCli/Program.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerCli/Program.cs:36]: added

**Review by asafMasa — APPROVED** [review]

**Review by shimoncohen — APPROVED** [review]

**Pull request merged**